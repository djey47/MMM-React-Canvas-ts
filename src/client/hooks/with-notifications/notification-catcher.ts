export interface NotificationCatcherOptions {
  isDebugMode?: boolean;
}

/**
 * Catcher singleton to be module-wide used
 *
 * It must be instantiated in an early module lifecycle step.
 * Its goal is to relay notifications coming from related lifecycle functions.
 * This will provide a way to plug React components (via HOC) to (socket/non socket) notifications.
 */
export class NotificationCatcher {
  static instance?: NotificationCatcher = undefined;
  notificationHandler: (n: string, p?: unknown) => void;
  subscribedNotifications: string[];
  isInitialized: boolean;
  options: NotificationCatcherOptions;

  constructor(options?: NotificationCatcherOptions) {
    this.notificationHandler = this.defaultHandler;
    this.subscribedNotifications = ['*'];
    this.isInitialized = false;
    this.options = {
      isDebugMode: false,
      ...options,
    };

    Log.log(
      '**** with-notifications: Notification catcher ready for initialization'
    );
  }

  /**
   * @returns the one and only notification catcher instance
   */
  public static getInstance(options?: NotificationCatcherOptions) {
    if (!NotificationCatcher.instance) {
      NotificationCatcher.instance = new NotificationCatcher();
    }

    // Update options if necessary
    NotificationCatcher.instance.options = {
      ...NotificationCatcher.instance.options,
      ...options,
    };

    return NotificationCatcher.instance;
  }

  /**
   * Makes catcher instance process notifications with a custom handler
   * @param handler relay function to be invoked on subscribed notification received
   * @param subscribed list of notification codes to store and relay payload data from ('*' will relay everything)
   */
  public switchHandler(
    handler: (n: string, p?: unknown) => void,
    subscribed: string[]
  ) {
    if (
      handler.toString() === this.notificationHandler.toString() &&
      subscribed === this.subscribedNotifications
    ) {
      return;
    }

    this.notificationHandler = handler;
    this.subscribedNotifications = subscribed;

    this.debugLog(
      `Notification catcher switched to handler ${handler} for ${subscribed}`
    );
  }

  /**
   * Passes a notification to the catcher if satisfies subscription
   *
   * To be used typically in MM2 lifecycle functions
   *
   * @param notif notification code
   * @param payload notification contents, eventually
   */
  public catchNotification(notif: string, payload?: unknown) {
    if (
      this.subscribedNotifications.includes('*') ||
      this.subscribedNotifications.includes(notif)
    ) {
      this.debugLog(
        `Notification catcher relaying notification {${notif}:${payload}}`
      );

      this.notificationHandler(notif, payload);
    } else {
      this.debugLog(
        `Notification catcher rejecting notification without subscription {${notif}:${payload}}`
      );
    }
  }

  private defaultHandler(n: string, p?: unknown) {
    this.debugLog(
      `Notification catcher processing notification with default handler {${n}:${p}}`
    );
  }

  private debugLog(message: string) {
    if (this.options.isDebugMode) {
      Log.log(`**** with-notifications: ${message}`);
    }
  }
}
