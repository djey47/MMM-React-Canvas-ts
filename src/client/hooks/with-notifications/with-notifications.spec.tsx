/* eslint-disable @typescript-eslint/ban-ts-comment */

import { FunctionComponent } from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';
import { NotificationCatcher } from './notification-catcher';
import useWithNotifications from './with-notifications';

describe('with-notifications hook', () => {
  beforeAll(() => {
    // @ts-ignore
    global.Log = {
      log: jest.fn(),
    };
  });

  describe('withNotifications function', () => {
    it('should do nothing without notification', () => {
      // given-when-then
      const tree = create(<TestComponent prop1="val1" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should provide data from notification', () => {
      // given
      let root: ReactTestRenderer;
      act(() => {
        root = create(<TestComponent prop1="val1" />);
      });

      // when
      act(() => {
        NotificationCatcher.getInstance().catchNotification('TOPIC', {
          data: 1,
        });
        root.update(<TestComponent prop1="val1" prop2="val2" />);
      });

      // then
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(root!.toJSON()).toMatchSnapshot();
    });

    it('should handle notification without data', () => {
      // given
      let root: ReactTestRenderer;
      act(() => {
        root = create(<TestComponent prop1="val1" />);
      });

      // when
      act(() => {
        NotificationCatcher.getInstance().catchNotification('TOPIC');
        root.update(<TestComponent prop1="val1" prop2="val2" />);
      });

      // then
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(root!.toJSON()).toMatchSnapshot();
    });
  });
});

interface TestComponentProps {
  prop1: string;
  prop2?: string;
}

const TestComponent: FunctionComponent<TestComponentProps> = (
  props: TestComponentProps
) => {
  const hookData = useWithNotifications(['TOPIC']);
  return <div {...props}>Test component: {JSON.stringify(hookData, null, 2)}</div>;
};
