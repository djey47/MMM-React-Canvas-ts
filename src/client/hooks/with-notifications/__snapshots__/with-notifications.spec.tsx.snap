// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with-notifications hook withNotifications function should do nothing without notification 1`] = `
<div
  prop1="val1"
>
  Test component: 
  {}
</div>
`;

exports[`with-notifications hook withNotifications function should handle notification without data 1`] = `
<div
  prop1="val1"
  prop2="val2"
>
  Test component: 
  {
  "data_TOPIC": {}
}
</div>
`;

exports[`with-notifications hook withNotifications function should provide data from notification 1`] = `
<div
  prop1="val1"
  prop2="val2"
>
  Test component: 
  {
  "data_TOPIC": {
    "data": 1
  }
}
</div>
`;
